### PR TITLE
Fix String value issues in grub_config

### LINKED ISSUE #44: grub_config values with spaces cause augeas errors
grub_config values with spaces cause augeus errors:

        grub_config {'GRUB_TERMINAL':
            value   => join($boot::grub_terminals, ' '),
            notify  => Exec['update-grub-config'],
        }

        grub_config {'GRUB_SERIAL_COMMAND':
            value   => "serial --unit=0 --speed=${boot::serialconsole}",
            notify  => Exec['update-grub-config'],
        }

Notice: /Stage[main]/Boot::Grub2/Grub_config[GRUB_TERMINAL]/value: value changed '"console serial"' to 'console serial'
Error: /Stage[main]/Boot::Grub2/Grub_config[GRUB_TERMINAL]: Could not evaluate: Failed to save Augeas tree to file. See debug logs for details.

Notice: /Stage[main]/Boot::Grub2/Grub_config[GRUB_SERIAL_COMMAND]/value: value changed '"serial --unit=0 --speed=115200"' to 'serial --unit=0 --speed=115200'
Error: /Stage[main]/Boot::Grub2/Grub_config[GRUB_SERIAL_COMMAND]: Could not evaluate: Failed to save Augeas tree to file. See debug logs for details.

### PR DESCRIPTION
**Pull request by trevor-vaughan**
* Ensure that Boolean values are converted to Strings
* Ensure the String values have quotes around them if not already
  present

Fixes #44, #45

### CODE REVIEW

**Comment by trevor-vaughan** [comment]
@raphink Up for a review?

File: lib/puppet/provider/grub_config/grub2.rb [hunk 1]
```diff
@@ -66,6 +66,12 @@ def value
   end
 
   def value=(newval)
+    if newval.is_a?(String)
+      unless %w[' "].include?(newval[0].chr)
+        newval = %Q("#{newval}")
```
Comment (raphink) [inline lib/puppet/provider/grub_config/grub2.rb]: I don't understand what you're doing here

File: lib/puppet/type/grub_config.rb [hunk 2]
```diff
@@ -45,6 +45,18 @@
     desc <<-EOM
       Value of the GRUB parameter.
     EOM
+
+    munge do |value|
+      value.to_s unless [Hash, Array].include?(value.class)
+    end
+
+    def insync?(is)
+      if is.is_a?(String) && should.is_a?(String)
```
Comment (raphink) [inline lib/puppet/type/grub_config.rb]: At some point, we should probably rewrite `augeasproviders` with the new Resource API to natively support data types.

File: lib/puppet/provider/grub_config/grub2.rb (hunk 1 above)
Comment (trevor-vaughan) [reply lib/puppet/provider/grub_config/grub2.rb]: @raphink Ensuring that we don't quote strings that already have quotes in them.

File: lib/puppet/type/grub_config.rb (hunk 2 above)
Comment (trevor-vaughan) [reply lib/puppet/type/grub_config.rb]: I'd be curious to see how that works out. From what I've seen, easy things are easy and advanced things are nigh impossible (might be missing something though).

File: lib/puppet/provider/grub_config/grub2.rb (hunk 1 above)
Comment (raphink) [reply lib/puppet/provider/grub_config/grub2.rb]: Oh I see, `%Q("")` adds literal quotes while interpreting `#{newval}`.

Comment (raphink) [reply lib/puppet/provider/grub_config/grub2.rb]: iirc, there's a a `quote_it` function in augeasproviders_core, which is used in the shellvar module.

File: lib/puppet/type/grub_config.rb (hunk 2 above)
Comment (raphink) [reply lib/puppet/type/grub_config.rb]: That was my feeling as well. @DavidS assured me otherwise though…

Comment (DavidS) [reply lib/puppet/type/grub_config.rb]: that severely depends on your definition of "advanced".

Comment (DavidS) [reply lib/puppet/type/grub_config.rb]: looking over the `grub_config` type and provider, this autorequire computation seems to be the one thing that the Resource API can't handle: https://github.com/hercules-team/augeasproviders_grub/blob/402ee21e8a590ed2c4e0c3981e35c7a03d268c61/lib/puppet/type/grub_config.rb#L63-L85

Comment (raphink) [reply lib/puppet/type/grub_config.rb]: The stuff we do in https://github.com/hercules-team/augeasproviders_core/blob/master/lib/puppet/provider/augeasprovider/default.rb for ex, such as the handler sharing in the Puppet session.

`instances` such as https://github.com/hercules-team/augeasproviders_ssh/blob/master/lib/puppet/provider/sshd_config/augeas.rb#L113

probably other things, too. @trevor-vaughan do you have things in mind?

Comment (DavidS) [reply lib/puppet/type/grub_config.rb]: The `instances` method seems to be fine - that's a regular `get`, collecting and computing all the resources.

I didn't fully read through the handler sharing, but you should be able to share any instances you need across providers by directly reaching into other ruby namespaces. No need to load something into the catalog for that.

**Comment by tparkercbn** [comment]
Hello.  I'm wondering when this will be merged to master and pushed to the forge?

Thank you

Tom

**Pull request merged**